crates/dojo-core/src/database/index.cairo

```diff
@@ -18,6 +18,48 @@ fn create(address_domain: u32, index: felt252, id: felt252) {
     storage::set(address_domain, build_index_key(index, index_len), id);
 }
 
+fn create_with_keys(address_domain: u32, index: felt252, id: felt252, keys: Span<felt252>) {
+    // TODO: handle reapeated
+    assert(keys.len() < 255, 'Too many keys');
+    create(address_domain, index, id);
+
+    let mut positions = ArrayTrait::<felt252>::new();
+    positions.append(keys.len().into());
+
+    let mut idx = 0;
+    loop {
+        if idx == keys.len() {
+            break ();
+        }
+        let pos = add_key(address_domain, index, id, *keys.at(idx), idx); // key -> id
+        positions.append(pos);
+        idx += 1;
+    };
+
+    let index_len_key = build_index_len_key(index);
+
+    let keys_len: u8 = keys.len().try_into().unwrap();
+    storage::set_many(address_domain, build_index_item_keys(index, id), 0, positions.span());  // len of keys and positions
+    storage::set_many(address_domain, build_index_item_keys(index, id), keys_len + 1, keys);   // keys
+}
+
+fn add_key(address_domain: u32, index: felt252, id: felt252, key: felt252, idx: u32) -> felt252 {
```

Comment:
just to make sure i understand, is `key` here the same as the dojo meaning of `key`? i.e. the unique id?

we want to support querying entities based on not only the key, but also attribute values.

Reply:
That's correct, we focused for now to produce some proof of concept for the indexing.
We already have an idea how to support indexing by attribute values.
Do we proceed indexing by attribute values in this PR as well?